tests/unit/test_probes_provider.py

```diff
@@ -0,0 +1,176 @@
+
+        self.harness.charm.provider.set_probes_spec()
+
+        data = self.harness.get_relation_data(rel_id, self.harness.model.app.name)
+        self.assertIn("scrape_probes", data)
+        scrape_data = json.loads(data["scrape_probes"])
+        self.assertEqual(scrape_data[0]["static_configs"][0]["targets"], ["10.1.238.1"])
+        self.assertEqual(scrape_data[0]["params"]["module"], ["http_2xx"])
+
+    def test_provider_sets_modules_with_prefix_on_relation_joined(self):
+        rel_id = self.harness.add_relation(RELATION_NAME, "provider")
+        self.harness.add_relation_unit(rel_id, "provider/0")
+
+        self.harness.charm.provider.set_probes_spec()
+
+        data = self.harness.get_relation_data(rel_id, self.harness.model.app.name)
+        self.assertIn("scrape_modules", data)
+
+        scrape_modules = json.loads(data["scrape_modules"])
+
+        topology = JujuTopology.from_dict(json.loads(data["scrape_metadata"]))
+        module_name_prefix = "juju_{}_".format(topology.identifier)
+
+        self.assertIn(f"{module_name_prefix}http_2xx_longer_timeout", scrape_modules)
+
+    def test_provider_prefixes_jobs(self):
+        rel_id = self.harness.add_relation(RELATION_NAME, "provider")
+        self.harness.add_relation_unit(rel_id, "provider/0")
+
+        self.harness.charm.provider.set_probes_spec()
+
+        data = self.harness.get_relation_data(rel_id, self.harness.model.app.name)
+        scrape_data = json.loads(data["scrape_probes"])
+        topology = JujuTopology.from_dict(json.loads(data["scrape_metadata"]))
+        module_name_prefix = "juju_{}_".format(topology.identifier)
+
+        self.assertEqual(scrape_data[0]["job_name"], f"{module_name_prefix}my-first-job")
+
+
+class BlackboxProbesProviderCharmWithWrongProbe(CharmBase):
+    _stored = StoredState()
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args)
+        self._stored.set_default(num_events=0)
+        self.provider = BlackboxProbesProvider(
+            self, probes=PROBES_NOT_VALID_MISSING_MODULE, modules=MODULES
+        )
+        self.framework.observe(self.provider.on.invalid_probe, self.record_events)
+
+    def record_events(self, event):
+        self._stored.num_events += 1
+
+
+class BlackboxProbesWrongProviderTest(unittest.TestCase):
+    def setUp(self):
+        self.harness = Harness(BlackboxProbesProviderCharmWithWrongProbe, meta=PROVIDER_META)
+        self.harness.set_model_name("MyUUID")
+        self.addCleanup(self.harness.cleanup)
+        self.harness.set_leader(True)
+        self.harness.begin()
+
+    def test_provider_notifies_on_invalid_probe(self):
+        self.assertEqual(self.harness.charm._stored.num_events, 0)
+        rel_id = self.harness.add_relation(RELATION_NAME, "provider")
+        self.harness.add_relation_unit(rel_id, "provider/0")
+        self.harness.charm.provider.set_probes_spec()
+        self.assertEqual(self.harness.charm._stored.num_events, 2)
```

Comment:
It's hard do understand the behavior being tested. Ideally there shouldn't be references to "num_events" in tests.

Reply:
This kind of whitebox testing tends to be brittle and difficult to maintain.
See BDD section [here](https://discourse.charmhub.io/t/best-practices-for-testing-charms/15172).